cmd/regctl/digest.go

```diff
@@ -30,18 +46,74 @@ func NewDigestCmd(rootOpts *rootCmd) *cobra.Command {
 		RunE:   digestOpts.runDigest,
 	}
 
+	digestCmd.Flags().StringVarP(&digestOpts.decompress, "decompress", "", "", "Decompress the bytes, using one of 'auto', 'zstd', 'gzip'")
 	return digestCmd
 }
 
 func (digestOpts *digestCmd) runDigest(cmd *cobra.Command, args []string) error {
+	var reader io.Reader = bufio.NewReader(cmd.InOrStdin())
+
+	var decompress compression
+	var err error
+
+	// Detect the compression in stdin
+	if digestOpts.decompress == "auto" {
+		decompress, err = getCompression(reader.(*bufio.Reader))
+		if err != nil {
+			return err
+		}
+	} else if digestOpts.decompress == "gzip" {
+		decompress = compressionGzip
+	} else if digestOpts.decompress == "zstd" {
+		decompress = compressionZstd
+	}
+
+	if decompress == compressionGzip {
+		reader, err = gzip.NewReader(reader)
+		if err != nil {
+			return err
+		}
+	} else if decompress == compressionZstd {
+		reader = zstd.ZipDecompressor()(reader)
+	}
+
 	digester := digest.Canonical.Digester()
 
-	_, err := io.Copy(digester.Hash(), os.Stdin)
+	_, err = io.Copy(digester.Hash(), reader)
 
 	if err != nil {
 		return err
 	}
 
-	fmt.Println(digester.Digest().String())
+	fmt.Fprintln(cmd.OutOrStdout(), digester.Digest().String())
 	return nil
 }
+
+func getCompression(reader *bufio.Reader) (compression, error) {
+	isZstd, err := hasMagicBytes(reader, zstdMagicHeader)
+	if err != nil {
+		return compressionNone, err
+	} else if isZstd {
+		return compressionZstd, nil
+	}
+
+	isGzip, err := hasMagicBytes(reader, gzipMagicHeader)
+	if err != nil {
+		return compressionNone, err
+	} else if isGzip {
+		return compressionGzip, nil
+	}
+
+	return compressionNone, nil
+}
```

Comment:
I've got this code already in pkg/archive. Lets import that and add zstd support there.

Reply:
Did not see that, removed everything here, and switched to archive pkg.